crates/typst/src/geom/color.rs

```diff
@@ -334,183 +954,696 @@ cast! {
+    }
+
+    fn alpha(self) -> Option<f64> {
+        Some(self.a)
+    }
+
+    fn to_rgba(self) -> RgbaColor {
+        Rgba::from_color(LinearRgba::from(self)).into()
+    }
+
+    fn to_oklab(self) -> OklabColor {
+        Oklab::from_color(LinearRgba::from(self)).into()
+    }
+
+    fn to_linear_rgb(self) -> LinearRgbColor {
+        self
+    }
+
+    fn to_hsl(self) -> HslColor {
+        self.to_rgba().to_hsl()
+    }
+
+    fn to_hsv(self) -> HsvColor {
+        self.to_rgba().to_hsv()
+    }
+
+    fn to_cmyk(self) -> CmykColor {
+        self.to_rgba().to_cmyk()
+    }
+
+    fn to_luma(self) -> LumaColor {
+        Luma::from_color(LinearRgba::from(self)).into()
+    }
+
+    fn lighten(self, factor: Ratio) -> Self {
+        LinearRgba::from(self).lighten(factor.get()).into()
+    }
+
+    fn darken(self, factor: Ratio) -> Self {
+        LinearRgba::from(self).darken(factor.get()).into()
+    }
+
+    fn saturate(self, factor: Ratio) -> Self {
+        self.to_hsv().saturate(factor).to_linear_rgb()
+    }
+
+    fn desaturate(self, factor: Ratio) -> Self {
+        self.to_hsv().desaturate(factor).to_linear_rgb()
+    }
+
+    fn hue_rotate(self, hue: Angle) -> Self {
+        self.to_hsv().hue_rotate(hue).to_linear_rgb()
+    }
+
+    fn negate(self) -> Self {
+        Self::new(1.0 - self.r, 1.0 - self.g, 1.0 - self.b, self.a)
+    }
+
+    fn to_array(self, alpha: bool) -> Array {
+        if alpha {
+            array![
+                Ratio::new(self.r),
+                Ratio::new(self.g),
+                Ratio::new(self.b),
+                Ratio::new(self.a),
+            ]
+        } else {
+            array![Ratio::new(self.r), Ratio::new(self.g), Ratio::new(self.b),]
+        }
+    }
+}
+
+impl LinearRgbColor {
+    pub fn new(r: f64, g: f64, b: f64, a: f64) -> Self {
+        Self { r, g, b, a }
+    }
+}
+
+impl Eq for LinearRgbColor {}
+
+impl Hash for LinearRgbColor {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.r.to_bits().hash(state);
+        self.g.to_bits().hash(state);
+        self.b.to_bits().hash(state);
+        self.a.to_bits().hash(state);
+    }
+}
+
+impl Debug for LinearRgbColor {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        if self.a == 1.0 {
+            write!(
+                f,
+                "linear-rgb({:?}, {:?}, {:?})",
+                Ratio::new(self.r),
+                Ratio::new(self.g),
+                Ratio::new(self.b),
+            )
+        } else {
+            write!(
+                f,
+                "linear-rgb({:?}, {:?}, {:?}, {:?})",
+                Ratio::new(self.r),
+                Ratio::new(self.g),
+                Ratio::new(self.b),
+                Ratio::new(self.a),
+            )
+        }
+    }
+}
+
+impl From<LinearRgbColor> for Color {
+    fn from(linear_srgb: LinearRgbColor) -> Self {
+        Self::LinearRgb(linear_srgb)
+    }
+}
+
+/// An 8-bit RGBA color.
+#[derive(Copy, Clone)]
+pub enum RgbaColor {
```

Comment:
Why does RGBA support both things? Why not always float? Also, the comment above is out of date with 8 bit.